### PR TITLE
Add Other type colorization & Fixes

### PR DESCRIPTION
**Pull request by DmitrySvetlichny**

### CODE REVIEW

File: Sources/Controllers/DashboardOnMap/MapSettings/OAMapSettingsTerrainParametersViewController.mm [hunk 1]
```diff
@@ -386,12 +412,30 @@ - (void)resetGPXVerticalExaggerationValues
     }
 }
 
+- (void)resetGPXElevationMetersValues
+{
+    NSInteger defaultMeters = 1000;
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/OAMapSettingsTerrainParametersViewController.mm:417]: kElevationDefMeters

**Review by tigrim — COMMENTED** [review]
Also, make sure not to write any 3D visualization data to the file if 'none' is selected. Export

File: Sources/Controllers/DashboardOnMap/MapSettings/OAMapSettingsTerrainParametersViewController.mm [hunk 2]
```diff
@@ -109,6 +112,15 @@ - (void)configureGPXVerticalExaggerationScale:(CGFloat)scale
     }
 }
 
+- (void)configureGPXElevationMeters:(NSInteger)meters
```
[the remark this answers is not in the document]
Comment (DmitrySvetlichny) [reply Sources/Controllers/DashboardOnMap/MapSettings/OAMapSettingsTerrainParametersViewController.mm:115]: always in meters

File: Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuAppearanceHudViewController.mm [hunk 3]
```diff
@@ -1524,15 +1626,13 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
...
             _trackView3DEmptyView.translatesAutoresizingMaskIntoConstraints = NO;
-            
-            [cell.contentView addSubview:_trackView3DEmptyView];
+
             [NSLayoutConstraint activateConstraints:@[
                 [_trackView3DEmptyView.leadingAnchor constraintEqualToAnchor:cell.contentView.leadingAnchor constant:20],
                 [_trackView3DEmptyView.trailingAnchor constraintEqualToAnchor:cell.contentView.trailingAnchor constant:-20],
                 [_trackView3DEmptyView.topAnchor constraintEqualToAnchor:cell.contentView.topAnchor constant:20],
-                [_trackView3DEmptyView.bottomAnchor constraintEqualToAnchor:cell.contentView.bottomAnchor constant:20],
+                [_trackView3DEmptyView.bottomAnchor constraintEqualToAnchor:cell.contentView.bottomAnchor constant:-20],
```
[the remark this answers is not in the document]
Comment (DmitrySvetlichny) [reply Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuAppearanceHudViewController.mm:1635]: Yes

File: Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuAppearanceHudViewController.mm [hunk 4]
```diff
@@ -845,25 +924,46 @@ - (void)generateData
...
-                    @"accessibility_label":OALocalizedString(@"vertical_exaggeration"),
-                    @"accessibility_value":alphaValueString,
-                    @"accessoryType":@(UITableViewCellAccessoryDisclosureIndicator)
-                },
-            }];
-            [track3DSectionItems addObject:verticalExaggerationData];
-        }
+                if (self.gpx.visualization3dByType != EOAGPX3DLineVisualizationByTypeFixedHeight)
+                {
+                    OAGPXTableCellData *verticalExaggerationData = [OAGPXTableCellData withData:@{
```
[the remark this answers is not in the document]
Comment (DmitrySvetlichny) [reply Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuAppearanceHudViewController.mm:955]: Fixed

File: Sources/Controllers/DashboardOnMap/MapSettings/OAMapSettingsTerrainParametersViewController.mm [hunk 5]
```diff
@@ -220,6 +240,14 @@ - (void)generateData
             kCellTitleKey : OALocalizedString(@"shared_string_scale")
         }];
     }
+    else if (_terrainType == EOAGPXSettingsTypeWallHeight)
+    {
+        [topSection addRowFromDictionary:@{
+            kCellKeyKey : @"wallHeightSlider",
+            kCellTypeKey : [OATitleSliderTableViewCell getCellIdentifier],
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/OAMapSettingsTerrainParametersViewController.mm:247]: already you can use reuseIdentifier

File: Sources/Controllers/Map/Layers/OAGPXLayer.mm [hunk 6]
```diff
@@ -411,6 +467,83 @@ - (void) refreshGpxTracks
...
+    OAGpxExtension *trackpointextension = [point getExtensionByKey:@"trackpointextension"];
+    NSNumber *waterTempValue = nil;
+    NSNumber *airTempValue = nil;
+    
+    if (trackpointextension)
+    {
+        for (OAGpxExtension *subextension in trackpointextension.subextensions)
+        {
+            if ([subextension.name isEqualToString:OAPointAttributes.sensorTagTemperatureW])
+                waterTempValue = [numberFormatter numberFromString:subextension.value];
```
Comment (tigrim) [inline Sources/Controllers/Map/Layers/OAGPXLayer.mm:533]: add return waterTempValue

File: Sources/Controllers/Map/Layers/OAGPXLayer.mm [hunk 7]
```diff
@@ -411,6 +467,83 @@ - (void) refreshGpxTracks
...
+            else if ([subextension.name isEqualToString:OAPointAttributes.sensorTagTemperatureA])
+                airTempValue = [numberFormatter numberFromString:subextension.value];
```
Comment (tigrim) [inline Sources/Controllers/Map/Layers/OAGPXLayer.mm:535]: add return airTempValue

File: Sources/Controllers/Map/Layers/OAGPXLayer.mm [hunk 8]
```diff
@@ -591,7 +724,13 @@ - (void)refreshLine:(QVector<OsmAnd::PointI> &)points
 {
     auto traceColorizationMapping = QList<OsmAnd::FColorARGB>();
     BOOL showTransparentTraces = gpx.visualization3dWallColorType == EOAGPX3DLineVisualizationWallColorTypeDownwardGradient
```
Comment (tigrim) [inline Sources/Controllers/Map/Layers/OAGPXLayer.mm:726]: add method shouldShowTransparentTraces

File: Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuAppearanceHudViewController.mm [hunk 9]
```diff
@@ -2098,6 +2215,31 @@ - (void)onButtonPressed:(OAGPXBaseTableData *)tableData
...
+        [controller configureGPXElevationMeters:savedElevationMeters];
+        __weak __typeof(self) weakSelf = self;
+        controller.applyWallHeightCallback = ^(NSInteger meters)
+        {
+            [weakSelf configureElevationMeters:meters];
+        };
+        controller.hideCallback = ^{
+            OATrackMenuViewControllerState *state = _reopeningTrackMenuState;
+            BOOL openedFromTracksList = state.openedFromTracksList;
+            state.openedFromTracksList = openedFromTracksList;
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuAppearanceHudViewController.mm:2231]: state.openedFromTracksList = state.openedFromTracksList;

File: Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuAppearanceHudViewController.mm [hunk 10]
```diff
@@ -484,6 +491,10 @@ - (UIMenu *)createMenuForKey:(NSString *)key button:(UIButton *)button
 - (void) configureVisualization3dByType:(EOAGPX3DLineVisualizationByType)type
 {
     self.gpx.visualization3dByType = type;
+    
+    if (self.gpx.visualization3dByType == EOAGPX3DLineVisualizationByTypeFixedHeight)
+        self.gpx.verticalExaggerationScale = 0.25;
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuAppearanceHudViewController.mm:496]: const

File: Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuAppearanceHudViewController.mm [hunk 11]
```diff
@@ -578,84 +620,114 @@ - (void)reloadTableWithAnimation
...
+        NSString *title = OALocalizedString(visualizationTypes[type]);
+        NSString *dataKey = dataKeys[type];
+        if (dataKey && ![self hasValidDataForKey:dataKey])
+            continue;
+        
+        UIAction *action = [UIAction actionWithTitle:title image:nil identifier:nil handler:^(__kindof UIAction * _Nonnull action) {
+            [weakSelf configureVisualization3dByType:(EOAGPX3DLineVisualizationByType)type.intValue];
+        }];
+        
+        if (self.gpx.visualization3dByType == type.intValue)
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuAppearanceHudViewController.mm:658]: type.intValue to let

File: Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuAppearanceHudViewController.mm [hunk 12]
```diff
@@ -578,84 +620,114 @@ - (void)reloadTableWithAnimation
 
 - (UIMenu *)createVisualizedByMenuForCellButton:(UIButton *)button
 {
-    NSMutableArray<UIMenuElement *> *menuElements = [NSMutableArray array];
     __weak __typeof(self) weakSelf = self;
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuAppearanceHudViewController.mm:623]: move before for

File: Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuAppearanceHudViewController.mm [hunk 13]
```diff
@@ -843,27 +943,44 @@ - (void)generateData
             }];
                 [track3DSectionItems addObject:trackLineData];
                 
-            NSString *alphaValueString = OALocalizedString(@"shared_string_none");
-            double scaleValue = self.gpx.verticalExaggerationScale;
-            if (scaleValue > 1)
-            {
-                alphaValueString = [NSString stringWithFormat:@"x%.1f", scaleValue];
+                double scaleValue = self.gpx.verticalExaggerationScale;
+                NSString *alphaValueString = scaleValue <= 0.25 ? OALocalizedString(@"shared_string_none") : (scaleValue < 1.0 ? [NSString stringWithFormat:@"x%.2f", scaleValue] : [NSString stringWithFormat:@"x%.1f", scaleValue]);
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuAppearanceHudViewController.mm:947]: 0.25 to const

File: Sources/GPX/OAGPXDatabase.mm [hunk 14]
```diff
@@ -672,7 +718,8 @@ - (OAGPX *)generateGpxItem:(NSDictionary *)gpxData
     bounds.bottomRight = CLLocationCoordinate2DMake([gpxData[@"bottom_right_lat"] doubleValue], [gpxData[@"bottom_right_lon"] doubleValue]);
     gpx.bounds = bounds;
     gpx.visualization3dWallColorType = EOAGPX3DLineVisualizationWallColorTypeUpwardGradient;
-    gpx.verticalExaggerationScale = 1.0;
+    gpx.verticalExaggerationScale = 0.25;
```
Comment (tigrim) [inline Sources/GPX/OAGPXDatabase.mm:721]: to const

File: Sources/GPX/OAGPXDocument.mm [hunk 15]
```diff
@@ -244,19 +247,37 @@ - (CGFloat)getVerticalExaggerationScale
     OAGpxExtension *e = [self getExtensionByKey:@"vertical_exaggeration_scale"];
     if (e) {
         CGFloat value = [e.value floatValue];
-        if (value && value >= 1.0 && value <= 3.0)
+        if (value && value >= 0.25 && value <= 4.0)
```
Comment (tigrim) [inline Sources/GPX/OAGPXDocument.mm:250]: use const

File: Sources/Controllers/Map/Layers/OAGPXLayer.mm [hunk 16]
```diff
@@ -473,6 +431,47 @@ - (void) refreshGpxTracks
     [self refreshStartFinishPoints];
 }
 
+- (void)processGPXDataElements:(NSArray *)elements withGPX:(OAGPX *)gpx addToElevations:(NSMutableArray *)elevations
+{
+    for (id element in elements)
+    {
+        NSArray *points = @[];
+        if ([element isKindOfClass:[OATrack class]]) {
```
Comment (tigrim) [inline Sources/Controllers/Map/Layers/OAGPXLayer.mm:439]: move down {

**Review by tigrim — APPROVED** [review]

**Pull request merged**